File: sdks/go/sip/privacy.go
package sip

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"
	"time"

	"golang.org/x/crypto/chacha20poly1305"
)

// GenerateViewingKey generates a new viewing key for selective disclosure.
func GenerateViewingKey(label string) (*ViewingKey, error) {
	key := make([]byte, 32)
	_, err := rand.Read(key)
	if err != nil {
		return nil, fmt.Errorf("failed to generate viewing key: %w", err)
	}

	keyHash := sha256.Sum256(key)
	createdAt := time.Now().UnixMilli()

	return &ViewingKey{
		Key:       BytesToHex(key),
		KeyHash:   BytesToHex(keyHash[:]),
		CreatedAt: createdAt,
		Label:     label,
	}, nil
}

// DeriveViewingKeyHash derives the hash of a viewing key.
//
// This hash is used for indexing and verification without
// exposing the actual key.
func DeriveViewingKeyHash(viewingKey HexString) (HexString, error) {
	keyBytes, err := HexToBytes(viewingKey)
	if err != nil {
		return "", fmt.Errorf("invalid viewing key: %w", err)
	}

	hash := sha256.Sum256(keyBytes)
	return BytesToHex(hash[:]), nil
}

// EncryptForViewingKey encrypts data for viewing key holders.
//
// Uses XChaCha20-Poly1305 for authenticated encryption.
func EncryptForViewingKey(viewingKey HexString, plaintext []byte) (*EncryptedPayload, error) {
	keyBytes, err := HexToBytes(viewingKey)
	if err != nil {
		return nil, fmt.Errorf("invalid viewing key: %w", err)
	}

	if len(keyBytes) != 32 {
		return nil, errors.New("viewing key must be 32 bytes")
	}

	// Create XChaCha20-Poly1305 cipher
	aead, err := chacha20poly1305.NewX(keyBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to create cipher: %w", err)
	}

	// Generate random nonce (24 bytes for XChaCha20)
	nonce := make([]byte, chacha20poly1305.NonceSizeX)
	_, err = rand.Read(nonce)
	if err != nil {
		return nil, fmt.Errorf("failed to generate nonce: %w", err)
	}

	// Encrypt
	ciphertext := aead.Seal(nil, nonce, plaintext, nil)

	return &EncryptedPayload{
		Ciphertext: BytesToHex(ciphertext),
		Nonce:      BytesToHex(nonce),
	}, nil
}

// DecryptWithViewingKey decrypts data using a viewing key.
func DecryptWithViewingKey(viewingKey HexString, payload *EncryptedPayload) ([]byte, error) {
	keyBytes, err := HexToBytes(viewingKey)
	if err != nil {
		return nil, fmt.Errorf("invalid viewing key: %w", err)
	}

	nonceBytes, err := HexToBytes(payload.Nonce)
	if err != nil {
		return nil, fmt.Errorf("invalid nonce: %w", err)
	}

	ciphertextBytes, err := HexToBytes(payload.Ciphertext)
	if err != nil {
		return nil, fmt.Errorf("invalid ciphertext: %w", err)
	}

	if len(keyBytes) != 32 {
		return nil, errors.New("viewing key must be 32 bytes")
	}

	if len(nonceBytes) != chacha20poly1305.NonceSizeX {
		return nil, errors.New("invalid nonce length")
	}

	// Create XChaCha20-Poly1305 cipher
	aead, err := chacha20poly1305.NewX(keyBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to create cipher: %w", err)
	}

	// Decrypt
	plaintext, err := aead.Open(nil, nonceBytes, ciphertextBytes, nil)
	if err != nil {
		return nil, fmt.Errorf("decryption failed: %w", err)
	}

	return plaintext, nil
}

// ShouldEncrypt determines if encryption should be used for a privacy level.
func ShouldEncrypt(level PrivacyLevel) bool {
	return level == PrivacyShielded || level == PrivacyCompliant
}

// ShouldIncludeViewingKey determines if viewing key should be included for a privacy level.
func ShouldIncludeViewingKey(level PrivacyLevel) bool {
	return level == PrivacyCompliant
}

File: sdks/rust/src/privacy.rs
//! Privacy and Viewing Key Implementation for SIP Protocol.
//!
//! Provides:
//! - Viewing key generation and derivation
//! - XChaCha20-Poly1305 encryption/decryption
//! - Selective disclosure for compliance

use chacha20poly1305::{
    aead::{Aead, KeyInit},
    XChaCha20Poly1305, XNonce,
};
use rand::RngCore;
use sha2::{Digest, Sha256};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::crypto::{bytes_to_hex, hex_to_bytes};
use crate::error::{Error, Result};
use crate::types::{EncryptedPayload, HexString, ViewingKey};

pub use crate::types::PrivacyLevel;

/// Generate a new viewing key for selective disclosure.
///
/// # Arguments
///
/// * `label` - Optional human-readable label
///
/// # Returns
///
/// ViewingKey object with key and hash
///
/// # Example
///
/// ```rust
/// use sip_protocol::generate_viewing_key;
///
/// let vk = generate_viewing_key(Some("audit-2024"));
/// ```
pub fn generate_viewing_key(label: Option<&str>) -> ViewingKey {
    let mut key = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut key);

    let mut hasher = Sha256::new();
    hasher.update(&key);
    let key_hash = hasher.finalize();

    let created_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64;

    ViewingKey {
        key: bytes_to_hex(&key),
        key_hash: bytes_to_hex(&key_hash),
        created_at,
        label: label.map(String::from),
    }
}

/// Derive the hash of a viewing key.
///
/// This hash is used for indexing and verification without
/// exposing the actual key.
pub fn derive_viewing_key_hash(viewing_key: &str) -> Result<HexString> {
    let key_bytes = hex_to_bytes(viewing_key)?;

    let mut hasher = Sha256::new();
    hasher.update(&key_bytes);
    let hash = hasher.finalize();

    Ok(bytes_to_hex(&hash))
}

/// Encrypt data for viewing key holders.
///
/// Uses XChaCha20-Poly1305 for authenticated encryption.
///
/// # Arguments
///
/// * `viewing_key` - The viewing key (32 bytes)
/// * `plaintext` - Data to encrypt
///
/// # Returns
///
/// EncryptedPayload with ciphertext and nonce
pub fn encrypt_for_viewing_key(viewing_key: &str, plaintext: &[u8]) -> Result<EncryptedPayload> {
    let key_bytes = hex_to_bytes(viewing_key)?;

    if key_bytes.len() != 32 {
        return Err(Error::CryptoError("Viewing key must be 32 bytes".to_string()));
    }

    // Generate random nonce (24 bytes for XChaCha20)
    let mut nonce_bytes = [0u8; 24];
    rand::thread_rng().fill_bytes(&mut nonce_bytes);
    let nonce = XNonce::from_slice(&nonce_bytes);

    // Create cipher and encrypt
    let cipher = XChaCha20Poly1305::new_from_slice(&key_bytes)
        .map_err(|e| Error::EncryptionError(e.to_string()))?;

    let ciphertext = cipher
        .encrypt(nonce, plaintext)
        .map_err(|e| Error::EncryptionError(e.to_string()))?;

    Ok(EncryptedPayload {
        ciphertext: bytes_to_hex(&ciphertext),
        nonce: bytes_to_hex(&nonce_bytes),
    })
}

/// Decrypt data using a viewing key.
///
/// # Arguments
///
/// * `viewing_key` - The viewing key (32 bytes)
/// * `payload` - The encrypted payload (ciphertext + nonce)
///
/// # Returns
///
/// Decrypted plaintext
pub fn decrypt_with_viewing_key(viewing_key: &str, payload: &EncryptedPayload) -> Result<Vec<u8>> {
    let key_bytes = hex_to_bytes(viewing_key)?;
    let nonce_bytes = hex_to_bytes(&payload.nonce)?;
    let ciphertext = hex_to_bytes(&payload.ciphertext)?;

    if key_bytes.len() != 32 {
        return Err(Error::CryptoError("Viewing key must be 32 bytes".to_string()));
    }

    if nonce_bytes.len() != 24 {
        return Err(Error::DecryptionError("Invalid nonce length".to_string()));
    }

    let nonce = XNonce::from_slice(&nonce_bytes);

    let cipher = XChaCha20Poly1305::new_from_slice(&key_bytes)
        .map_err(|e| Error::DecryptionError(e.to_string()))?;

    cipher
        .decrypt(nonce, ciphertext.as_slice())
        .map_err(|e| Error::DecryptionError(e.to_string()))
}

/// Determine if encryption should be used for a privacy level.
pub fn should_encrypt(level: PrivacyLevel) -> bool {
    matches!(level, PrivacyLevel::Shielded | PrivacyLevel::Compliant)
}

/// Determine if viewing key should be included for a privacy level.
pub fn should_include_viewing_key(level: PrivacyLevel) -> bool {
    matches!(level, PrivacyLevel::Compliant)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_viewing_key() {
        let vk = generate_viewing_key(Some("test"));
        assert!(vk.key.starts_with("0x"));
        assert!(vk.key_hash.starts_with("0x"));
        assert_eq!(vk.label, Some("test".to_string()));
    }

    #[test]
    fn test_encrypt_decrypt() {
        let vk = generate_viewing_key(None);
        let plaintext = b"Hello, SIP Protocol!";

        let payload = encrypt_for_viewing_key(&vk.key, plaintext).unwrap();
        let decrypted = decrypt_with_viewing_key(&vk.key, &payload).unwrap();

        assert_eq!(plaintext.as_slice(), decrypted.as_slice());
    }

    #[test]
    fn test_derive_viewing_key_hash() {
        let vk = generate_viewing_key(None);
        let hash = derive_viewing_key_hash(&vk.key).unwrap();
        assert_eq!(hash, vk.key_hash);
    }
}
